### PR TITLE
[Merged by Bors] - chore(Tactic/GCongr): move `@[gcongr]` tags around

### PR DESCRIPTION
**Pull request by urkud**
* Add `import Mathlib.Tactic.GCongr.Core`
  to `Algebra/Order/Ring/Lemmas`.
* Move most `@[gcongr]` tags next to the lemmas.

See [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Mathlib.2FTactic.2FGCongr)

---
- [x] Is there a better home for setting up `positivity`
  as a `gcongr` discharger?
- [x] What should we do with `Nat` division lemmas?

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/Field/Basic.lean
Comment (Parcly-Taxel) [inline Mathlib/Algebra/Order/Field/Basic.lean:350]: Yes, we do need it. I removed `@[gcongr]` from this lemma and L777 of `Analysis.Normed.Field.Basic` errored:
https://github.com/leanprover-community/mathlib4/blob/b50557ce70bf397971b2ba89d7c532c36a163e5f/Mathlib/Analysis/Normed/Field/Basic.lean#L771-L777
You never doubted that both `div_lt_div` and `div_lt_div_of_lt_left` should be gcongr lemmas, so if you never doubted `div_le_div`, how can `div_le_div_of_le_left` _not_ be a gcongr lemma without breaking something else in mathlib?

File: Mathlib/Tactic/GCongr.lean
Comment (Parcly-Taxel) [inline Mathlib/Tactic/GCongr.lean:29]: The three remaining gcongr lemmas in `Tactic.GCongr` can be removed **entirely** and mathlib still builds perfectly, see https://github.com/leanprover-community/mathlib4/commit/41f1b572c9ea37556e2252d9076cb51f1f5ee381

Comment (urkud) [reply Mathlib/Tactic/GCongr.lean:29]: But they should be tagged if we want `gcongr` to work with `Nat` division.

File: Mathlib/Algebra/Order/Field/Basic.lean
Comment (urkud) [reply Mathlib/Algebra/Order/Field/Basic.lean:350]:
```suggestion
@[gcongr]
```

File: Mathlib/Tactic/GCongr.lean
Comment (Parcly-Taxel) [reply Mathlib/Tactic/GCongr.lean:29]: My experiment showed that `gcongr` is not used to handle `Nat` division anywhere in the current library. But then of the three remaining `Nat` lemmas, only `Nat.div_le_div_right` sees individual use outside the place where it's declared, and only in three other files (besides `Tactic.GCongr` and the place where it's declared `Data.Nat.Basic`).

Thus I argue that there is no need – present or future – for `gcongr` to handle `Nat` division, so the tags should be removed (but `Nat.div_le_div` can be put into `Data.Nat.Basic`).

**Review by Parcly-Taxel — APPROVED** [review]
> Is there a better home for setting up positivity as a gcongr discharger?

I think the current setup is fine enough.

**Comment by urkud** [comment]
> > Is there a better home for setting up positivity as a gcongr discharger?
> 
> I think the current setup is fine enough.

It's a footgun: `gcongr` kind of works without importing `Mathlib.Tactic.GCongr` but doesn't use `positivity`. To reduce damage, we can `import` this file very early as soon as both `Positivity.Core` and `GCongr.Core` are imported. OTOH, we already have a similar footgun with `Positivity.Core` vs `Positivity`, so maybe the right way to solve the issue is to add documentation to `gcongr` and `positivity` ("if the tactic fails, first make sure that your file imports ...").

**Comment by Parcly-Taxel** [comment]
> so maybe the right way to solve the issue is to add documentation to `gcongr` and `positivity` ("if the tactic fails, first make sure that your file imports ...").

Just solve the issue as you see fit, then.

**Comment by Parcly-Taxel** [comment]
Do you still think there is a problem with the `positivity` import or is this ready?

File: Mathlib/Algebra/Order/Ring/Lemmas.lean
Comment (j-loreaux) [inline Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: Is there a reason not to simply import `Mathlib.Tactic.GCongr` here? That is, do we really not want to import `Mathlib.Tactic.Positivity`(`.Core`) here?

**Review by j-loreaux — COMMENTED** [review]
This basically looks fine to me. A few small issues:

1. Not for this PR, but the fact that `Algebra.Order.Ring.Defs` imports the `.Lemmas` file is weird. I opened a [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Lemmas.20vs.E2.80.8C.20Defs/near/416948726).
2. what are the `tt - ff` comments in the `GCongr` file, which are being deleted in this PR. Are we losing information of some sort?
3. It would be nice to have documented somewhere an example of a "bad" `gcongr` lemmma (i.e., one that a user might at first glance think would be a good lemma, but actually isn't for some reason). I don't see this currently covered.

File: Mathlib/Algebra/Order/Ring/Lemmas.lean
Comment (urkud) [reply Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: I'm good with importing positivity as early as possible. Should I ask on Zulip for other opinions?

Comment (urkud) [reply Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: I see that @semorrison recently removed some imports of `positivity`. Scott, what do you think about this?

**Comment by hrmacbeth** [comment]
LGTM, modulo: did this get resolved?

> I'm good with importing positivity as early as possible. Should I ask on Zulip for other opinions?

File: Mathlib/Algebra/Order/Ring/Lemmas.lean
Comment (semorrison) [reply Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: @j-loreaux, it would cause an import cycle!

Comment (urkud) [reply Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: So, should I merge `master`, wait for CI and merge?

Comment (Parcly-Taxel) [reply Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: Yes, go ahead and do that.

Comment (urkud) [reply Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: > Yes, go ahead and do that.

I'm waiting for an official `bors d+` from another maintainer.

Comment (Parcly-Taxel) [reply Mathlib/Algebra/Order/Ring/Lemmas.lean:8]: There's #10140 in the mix now. Would you like to adapt your PR to that? (cc @YaelDillies)

**Comment by YaelDillies** [comment]
Yes, negatively. `positivity` can't be imported earlier than `Field` because of the `norm_num` support at the core of the tactic. So #10140 only moves the extensions that depend on `Field` anyway.

**Comment by Parcly-Taxel** [comment]
#10140 has been merged.

**Comment by urkud** [comment]
Merged `master`, fixed compile.

File: Mathlib/Algebra/Order/Field/Basic.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Field/Basic.lean:351]: Does the same concern not apply to `gcongr`?

File: Mathlib/Algebra/Order/Field/Power.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Field/Power.lean:84]:
```suggestion
@[gcongr] alias ⟨_, GCongr.zpow_lt_of_lt⟩ := zpow_lt_iff_lt
```

File: Mathlib/Data/Nat/Cast/Order.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Cast/Order.lean:38]:
```suggestion
theorem GCongr.cast_le_cast {a b : ℕ} (h : a ≤ b) : (a : α) ≤ b := mono_cast h
```

Comment (j-loreaux) [reply Mathlib/Data/Nat/Cast/Order.lean:38]: Same comment as above.

Comment (YaelDillies) [reply Mathlib/Data/Nat/Cast/Order.lean:38]: This is a lemma internal to `GCongr`. If we don't namespace it, people, will be confused between `Nat.cast_le`, `Nat.cast_le_cast` and `Nat.cast_mono`. Further, the argument that we don't have much in the `GCongr` namespace doesn't hold since `gcongr` is so new.

File: Mathlib/Algebra/Order/Field/Basic.lean
Comment (urkud) [reply Mathlib/Algebra/Order/Field/Basic.lean:351]: It doesn't apply; `gcongr` wants left/right lemmas even if the "both arguments" lemma implies both of them.

File: Mathlib/Data/Nat/Cast/Order.lean
Comment (urkud) [reply Mathlib/Data/Nat/Cast/Order.lean:38]: This lemma was (not-namespaced) in `Tactic/GCongr`. We can move it to the `GCongr` namespace and deprecate the old lemma in favor of `Nat.mono_cast`.

Comment (YaelDillies) [reply Mathlib/Data/Nat/Cast/Order.lean:38]: I would consider the original lack of namespacing as an accident. I think we should not even deprecate but delete, since `Nat.cast_le` should really be called `Nat.cast_le_cast` and the current `Nat.cast_le_cast` hasn't been around for long.

File: Mathlib/Algebra/Order/Field/Power.lean
Comment (urkud) [reply Mathlib/Algebra/Order/Field/Power.lean:84]: I like the idea of moving auxiliary lemmas that exist only because `gcongr` can't handle `Monotone`/`StrictMono`/`Iff` lemmas to the `GCongr` namespace.

File: Mathlib/Data/Nat/Cast/Order.lean
Comment (urkud) [reply Mathlib/Data/Nat/Cast/Order.lean:38]: I deprecated old lemmas.

File: Mathlib/Algebra/Order/Field/Power.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Field/Power.lean:86]: to follow ISO recommendations
```suggestion
@[deprecated] -- Since 2024-02-10
```

File: Mathlib/Data/Nat/Cast/Order.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Cast/Order.lean:40]: Probably you meant this to go on the `@[deprecated]` above?

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Review by Vierkantor — APPROVED** [review]
LGTM, thanks!

bors r+

**Pull request closed without merging**
